src/ApiSchema.ts

```diff
@@ -1,46 +1,54 @@
 import Discord = require("discord.js");
 import fetch from "node-fetch";
+import XRegExp = require("xregexp");
 
 import levenshteinDistance from "./LevenshteinDistance";
 import { SharedSettings } from "./SharedSettings";
 
 import { clearTimeout, setTimeout } from "timers";
 
-class PathRegexCollection {
-    public invalid: string;
-    public valid: string;
-
-    constructor(validString: string, invalidString: string) {
-        this.valid = validString;
-        this.invalid = invalidString;
-    }
+export class Path {
+    public name: string;
+    public methodType: "GET" | "POST";
+    public regex: RegExp;
+    public pathParameters: Map<string, Parameter>;
+    public queryParameters: Map<string, Parameter>;
 }
 
-class SchemaRegexCollection extends PathRegexCollection {
+export class Parameter {
     public name: string;
-    public schema: any;
+    public required: boolean;
+    public type: ParameterType;
+}
 
-    constructor(name: string, schema: any, validString: string, invalidString: string) {
-        super(validString, invalidString);
+export class ParameterType {
+    /** A human readable description (e.g. "a positive integer") */
+    public description: string;
 
-        this.name = name;
-        this.schema = schema;
+    constructor(description: string, isValidValue: (value: string) => boolean){
+        this.description = description;
+        this.isValidValue = isValidValue;
     }
-}
-
-export class Path {
-    public methodType: "GET" | "POST";
-    public regex: PathRegexCollection;
-    public parameterInfo: SchemaRegexCollection[];
-    public method: string;
-    public name: string;
+    /** A function that returns a boolean indicating if the specified value is a valid value for this type of parameter */
+    public isValidValue(value: string | string[]): boolean { return false; }
 }
 
 export class APISchema {
 
+    private static PARAMETER_TYPES = {
+        ANY: new ParameterType("anything", (value) => true),
+        STRING: new ParameterType("a string", (value) => !Array.isArray(value)),
+        INTEGER: new ParameterType("an integer", (value) => Number.isInteger(+value)),
+        BOOLEAN: new ParameterType("a boolean", (value) => value === "true" || value === "false"),
+        SET: new ParameterType("a set specified like `paramName=value1&paramName=value2`", (value) => {
```

Comment:
Technically you can just do value without thr braces. More a stylistic choice though. I think we mostly have it without braces in this codebase?